groups/reconcile.go

```diff
@@ -0,0 +1,397 @@
+	}
+	if len(g.Groups) == 0 {
+		log.Println("No groups found.")
+		return nil
+	}
+	for _, g := range g.Groups {
+		// Don't touch existing mailing lists, we should
+		// always prefix with "k8s-infra-"
+		if !strings.HasPrefix(g.Email, "k8s-infra-") {
+			continue
+		}
+		log.Printf("%s\n", g.Email)
+
+		g2, err := srv2.Groups.Get(g.Email).Do()
+		if err != nil {
+			return fmt.Errorf("unable to retrieve group info for group %s: %v", g.Email, err)
+		}
+		log.Printf(">> Allow external members %s\n", g2.AllowExternalMembers)
+
+		l, err := srv.Members.List(g.Email).Do()
+		if err != nil {
+			return fmt.Errorf("unable to retrieve members in group : %v", err)
+		}
+
+		if len(l.Members) == 0 {
+			log.Println("No members found in group.")
+		} else {
+			for _, m := range l.Members {
+				log.Printf(">>> %s (%s)\n", m.Email, m.Role)
+			}
+		}
+		log.Printf("\n")
+
+	}
+	return nil
+}
+
+func createGroupIfNecessary(srv *admin.Service, groupEmailId string, description string) error {
+	_, err := srv.Groups.Get(groupEmailId).Do()
+	if err != nil {
+		if apierr, ok := err.(*googleapi.Error); ok && apierr.Code == http.StatusNotFound {
+			if config.DryRun {
+				log.Printf("dry-run : skipping creation of group %s\n", groupEmailId)
+			} else {
+				log.Printf("Trying to create group: %s\n", groupEmailId)
+				g4, err := srv.Groups.Insert(&admin.Group{
+					Email:       groupEmailId,
+					Name:        description,
+					Description: "Kubernetes wg-k8s-infra test group #2",
+				}).Do()
+				if err != nil {
+					return fmt.Errorf("unable to add new group %s: %v", groupEmailId, err)
+				}
+				log.Printf("> Successfully created group %s\n", g4.Email)
+			}
+		} else {
+			return fmt.Errorf("unable to fetch group: %#v", err.Error())
+		}
+	}
+	return nil
+}
+
+func deleteGroupsIfNecessary(service *admin.Service) error {
+	g, err := service.Groups.List().Customer("my_customer").OrderBy("email").Do()
+	if err != nil {
+		return fmt.Errorf("unable to retrieve users in domain: %v", err)
+	}
+	if len(g.Groups) == 0 {
+		log.Println("No groups found.")
+		return nil
+	}
+	for _, g := range g.Groups {
+		// Don't touch existing mailing lists, we should
+		// always prefix with "k8s-infra-"
+		if !strings.HasPrefix(g.Email, "k8s-infra-") {
+			continue
+		}
+		found := false
+		for _, g2 := range groupsConfig.Groups {
+			if g2.EmailId == g.Email {
+				found = true
+				break
+			}
+		}
+		if found {
+			continue
+		}
+		// We did not find the group in our groups.xml, so delete the group
+		if config.DryRun {
+			log.Printf("dry-run : Skipping removing group %s\n", g.Email)
+		} else {
+			err := service.Groups.Delete(g.Email).Do()
+			if err != nil {
+				return fmt.Errorf("unable to remove group %s : %v", g.Email, err)
+			}
+			log.Printf("Removing group %s\n", g.Email)
+		}
+
+	}
+	return nil
+}
+
+func updateGroupSettingsToAllowExternalMembers(srv *groupssettings.Service, groupEmailId string) error {
```

Comment:
Are there other settings we need to change here to make these lists secure?
https://developers.google.com/admin-sdk/groups-settings/v1/reference/groups#resource

I'm thinking things like `whoCanJoin`, `whoCanInvite, `whoCanAdd`, etc. And for the -private group, `whoCanViewGroup`.

Truly, I'd like to expose all of these settings via yaml, but we need a few of these at minimum.

Reply:
Added to follow up issue https://github.com/kubernetes/k8s.io/issues/255